File: 2018-01-1W/decomposition_reaction_dp.py
# problem: http://www.acmicpc.net/problem/2197
import sys

class DecomReaction:
    def read_input(self, input_method):
        self.N, self.M = [int(x) for x in input_method().split()]
        self.tree = [[] for _ in range(self.N)]
        self.subtree_count = [1 for _ in range(self.N)]

        self.min_cut_with_root = {}
        self.min_cut_without_root = {}

        for i in range(self.N-1):
            left_atom, right_atom = [int(x)-1 for x in input_method().split()]

            self.tree[left_atom].append(right_atom)
            self.tree[right_atom].append(left_atom)

    def count_subtree(self, node=None, visited=None):
        if node is None and visited is None:
            self.root = 0
            node = self.root
            visited = [False for _ in range(self.N)]

        visited[node] = True

        children = self.tree[node]
        new_children = []
        for child in children:
            if visited[child]:
                continue

            self.subtree_count[node] += self.count_subtree(child, visited)
            new_children.append(child)

        self.tree[node] = new_children

        return self.subtree_count[node]

    def solve_without_root(self, node, cut):
        if self.min_cut_without_root.get((node, cut), False):
            return self.min_cut_without_root[(node, cut)]

        count = self.subtree_count[node] 
        if count <= cut:
            self.min_cut_without_root[(node, cut)] = float('inf')
            return float('inf')

        ret = float('inf')
        children = self.tree[node]
        for child in children:
            ret = min(ret, 1 + self.solve_with_root(child, cut))
            ret = min(ret, self.solve_without_root(child, cut))

        self.min_cut_without_root[(node, cut)] = ret
        return ret

    def solve_with_root(self, node, cut):
        if self.min_cut_with_root.get((node, cut), False):
            return self.min_cut_with_root[(node, cut)]

        children = self.tree[node]
        if cut == 0:
            self.min_cut_with_root[(node, cut)] = 1
            return 1
        if cut == 1:
            self.min_cut_with_root[(node, cut)] = len(children)
            return len(children)

        count = self.subtree_count[node] 
        if count < cut:
            return float('inf')
        if count == cut:
            self.min_cut_with_root[(node, cut)] = 0
            return 0

        self.solve_all_combination(0, cut-1, 0, 0, node, cut)
        return self.min_cut_with_root[(node, cut)]

    def solve_all_combination(self, start, goal, progress, answer, root, cut):
        children = self.tree[root]
        if start == len(children)-1:
            old_answer = self.min_cut_with_root.get((root, cut), float('inf'))
            new_answer = answer + self.solve_with_root(children[start], goal - progress)
            self.min_cut_with_root[(root, cut)] = min(old_answer, new_answer)
            return

        for step in range(self.subtree_count[children[start]]+1):
            if progress + step > goal:
                continue
            delta = self.solve_with_root(children[start], step)
            self.solve_all_combination(start+1, goal, progress + step, answer + delta, root, cut)

    def solve(self):
        self.count_subtree()
        return min(self.solve_with_root(self.root, self.M), self.solve_without_root(self.root, self.M))


def main():
    dr = DecomReaction()
    dr.read_input(sys.stdin.readline)
    answer = dr.solve()
    print(answer)

if __name__ == '__main__':
    main()

File: 2018-01-1W/decomposition_reaction_dp.cpp
#include <stdio.h>
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;


class DecomReaction {
    public:
        int N, M;

        vector<int>* tree;
        vector<int>* tree_old;
        int* subtree_count;

        int** min_cut_with_root, ** min_cut_without_root;

        static const int root = 0;
        static const int INFINITY = 150;

        void readInput() {
            scanf("%d %d", &N, &M);
            tree_old = new vector<int>[N];
            tree     = new vector<int>[N];

            int v, w;
            for (int i = 0; i < N-1; i++) {
                scanf("%d %d", &v, &w);
                tree_old[v-1].push_back(w-1);
                tree_old[w-1].push_back(v-1);
            }
        }

        int solve() {
            subtree_count = new int[N];
            bool* visited = new bool[N];
            countSubtree(root, visited);
            delete[] tree_old;
            delete[] visited;

            min_cut_with_root = new int*[N];
            min_cut_without_root = new int*[N];

            for (int i = 0; i < N; i++) {
                min_cut_with_root[i]    = new int[N];
                min_cut_without_root[i] = new int[N];

                for (int j = 0; j < N; j++) {
                    min_cut_with_root[i][j]    = INFINITY;
                    min_cut_without_root[i][j] = INFINITY;
                }
            }

            int answer = min(solve_with_root(root, M, -1), solve_without_root(root, M, -1));
            cout << answer << endl;
            return answer;
        }

        ~DecomReaction() {
            delete[] tree;
            delete[] subtree_count;

            for (int i = 0; i < N; i++) {
                delete[] min_cut_with_root[i];
                delete[] min_cut_without_root[i];
            }

            delete[] min_cut_with_root;
            delete[] min_cut_without_root;
        }

        void printCache(vector<int>* cache) {
            for (int n = 0; n < N; n++) {
                cout << "node " << n << ": ";
                printVector(cache[n]);
            }
        }

        void printVector(vector<int> children) {
            for (vector<int>::iterator child = children.begin(); child != children.end(); child++) {
                cout << " " << *child;
            }
            cout << endl;
        }

    private:
        int countSubtree(int node, bool* visited) {
            vector<int> children = tree_old[node];
            vector<int>* children_new = &tree[node];

            visited[node] = true;

            subtree_count[node] = 1;
            for (vector<int>::iterator child = children.begin(); child != children.end(); child++) {
                if (visited[*child]) continue;

                subtree_count[node] += countSubtree(*child, visited);
                children_new->push_back(*child);
            }

            return subtree_count[node];
        }

        int solve_with_root(int node, int cut, int parent) {
            if (min_cut_with_root[node][cut] != INFINITY)
                return min_cut_with_root[node][cut];

            vector<int> children = tree[node];
            if (cut == 0) {
                min_cut_with_root[node][cut] = 1;
                return 1;
            }
            if (cut == 1) {
                min_cut_with_root[node][cut] = children.size()-1;
                return min_cut_with_root[node][cut];
            }

            int count = subtree_count[node];
            if (count < cut) {
                return INFINITY;
            }
            if (count == cut) {
                min_cut_with_root[node][cut] = 0;
                return 0;
            }

            solve_all_combination(0, cut-1, 0, 0, node, cut, parent);
            return min_cut_with_root[node][cut];
        }

        void solve_all_combination(int start, int goal, int progress, int answer, int node, int cut, int parent) {
            vector<int> children = tree[node];
            if (children[start] == parent) {
                if (start == children.size()-1) return;
                else solve_all_combination(start+1, goal, progress, answer, node, cut, parent);
            }
            if (start == children.size()-1) {
                int old_answer = min_cut_with_root[node][cut];
                int new_answer = answer + solve_with_root(children[start], goal - progress, node);
                min_cut_with_root[node][cut] = min(old_answer, new_answer);
                return;
            }

            for (int step = 0; step <= subtree_count[children[start]]; step++) {
                if (progress + step > goal) continue;

                int delta = solve_with_root(children[start], step, node);
                solve_all_combination(start+1, goal, progress + step, answer + delta, node, cut, parent);
            }

        }

        int solve_without_root(int node, int cut, int parent) {
            if (min_cut_without_root[node][cut] != INFINITY)
                return min_cut_without_root[node][cut];

            int count = subtree_count[node];
            if (count <= cut) {
                return INFINITY;
            }

            int ret = INFINITY;
            vector<int> children = tree[node];
            for (vector<int>::iterator child = children.begin(); child != children.end(); child++) {
                if (*child == parent) continue;
                ret = min(ret, 1 + solve_with_root(*child, cut, node));
                ret = min(ret, solve_without_root(*child, cut, node));
            }

            min_cut_without_root[node][cut] = ret;
            return ret;
        }
};

int main() {
    DecomReaction dr = DecomReaction();
    dr.readInput();
    dr.solve();
    return 0;
}
